### PR TITLE
Parameters 'client_id' and 'response_type' not strictly required in O…

### LINKED ISSUE #14255: Exception due to requirement of "client_id" presence within Request Object
### Describe the bug

Issue Description:
===================

A thirdparty OIDC client registered in Keycloak is using the authorisation code grant flow.
The authorization request is build by this OIDC client and the `request_uri` request parameter is provided.

When Keycloak tries to handle the JWE Encrypted Request Object obtained through the "request_uri" request parameter, we can see that Keycloak verifies that the "client_id" claim is present within the decrypted request object and is the same as the "client_id" provided request parameter. Request Object provided by a thirdparty OIDC Client doesn't contain this claim as this is not a mandatory claim in the OIDC specifications.

The OIDC specification states that for the request to be a valid OAuth 2.0 Authorization Request, " values for the response_type and client_id parameters MUST be included using the OAuth 2.0 request syntax, since they are REQUIRED by OAuth 2.0. The values for these parameters MUST match those in the Request Object, if present."
[OIDC Specification](
https://openid.net/specs/openid-connect-core-1_0.html#RequestUriParameter
)

The "response_type" and "client_id" request parameters are indeed provided by this thirdparty OIDC client but are not present in the Request Object because these values are optional within it. So the validation against the "client_id" should be done only if the claim is provided in the Request Object only... and then the presence of this claim shouldn't be mandatory as today within Keycloak

**Auth Request Endpoint**
https://www.foo.com/authorize?response_type=code
&redirect_uri=https%3A%2F%2Frfoo.com&request_uri=https%3A%2F%2Ffoo.com/bar&scope=openid%2Bservices&client_id=aaa&state=bbb&nonce=ccc

**Decrypted Request Object**
```
{
"foo": "bar",
"baz": "some"
}
```

server.log
===========

```
org.keycloak.protocol.oidc.endpoints.request.AuthorizationEndpointRequestParserProcessor.parseRequest(AuthorizationEndpointRequestParserProcessor.java:96)
at org.keycloak.protocol.oidc.endpoints.AuthorizationEndpoint.process(AuthorizationEndpoint.java:132)
at org.keycloak.protocol.oidc.endpoints.AuthorizationEndpoint.buildGet(AuthorizationEndpoint.java:112)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
```

Notes:
=====

The change should be done in the following class
https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestObjectParser.java
In particular...

```
if (clientId == null) {
throw new RuntimeException("Request object must be set with the client_id");
}

if (!client.getClientId().equals(clientId.asText())) {
throw new RuntimeException("The client_id in the request object is not the same as the authorizing client");
}
```

should be changed into

```
if (clientId != null && !client.getClientId().equals(clientId.asText())) {
throw new RuntimeException("The client_id in the request object is not the same as the authorizing client");
}
```

Workaround
===========
Updated source file of AuthzEndpointRequestObjectParser
Updated source file of AuthzEndpointRequestParser
https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java
 
https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestObjectParser.java

Queries:
=========
1) Is the "client_id" claim mandatory in the object request (current Keycloak mode) ?

2) Is the overriding of each authorization endpoint request attribute, with the one available from the one in the object request even if it is null, correct (current Keycloak mode) ?

3) Regarding the RFC9101 ("The OAuth 2.0 Authorization Framework: JWT-Secured Authorization Request (JAR)"), the "client_id" claim seems indeed mandatory.
However, it doesn't seem correct to override the authorization endpoint request attributes with null values if not present in the request object.

### Version

19.0.1

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mposolda**
…IDC request object

Closes #14255

### CODE REVIEW

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/OIDCAdvancedRequestParamsTest.java [hunk 1]
```diff
@@ -631,6 +631,56 @@ public void requestObjectSupersedesQueryParameter() throws Exception {
...
+        // Test that "client_id" mandatory in the query even if set in the "request" object
+        TestOIDCEndpointsApplicationResource oidcClientEndpointsResource = testingClient.testApp().oidcClientEndpoints();
+        oidcClientEndpointsResource.setOIDCRequest("test", "test-app", oauth.getRedirectUri(), "10", "some-state", "none");
+        oauth.request(oidcClientEndpointsResource.getOIDCRequest());
+        oauth.clientId(null);
+        oauth.openLoginForm();
+        errorPage.assertCurrent();
+
+        // Test that "response_type" mandatory in the query even if set in the "request" object
+        oidcClientEndpointsResource.setOIDCRequest("test", "test-app", oauth.getRedirectUri(), "10", "some-state", "none");
```
Comment (tnorimat) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/OIDCAdvancedRequestParamsTest.java:647]: This line and the following line might not be needed.

**Review by tnorimat — COMMENTED** [review]
@mposolda I added one review comment. Could you check it?

**Comment by tnorimat** [comment]
@mposolda I ran FAPI1 advanced final conformance test against this PR and confirmed that the PR can pass all conformance tests. I will run OIDC conformance tests and others against the PR.

**Comment by tnorimat** [comment]
@mposolda I ran other conformance tests (OIDC, OIDC for Logout Profiles, FAPI-CIBA, Open Banking Brazil FAPI ID2, Australia CDR, UK Open Banking) and confirmed that the PR can pass all conformance tests.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/OIDCAdvancedRequestParamsTest.java (hunk 1 above)
Comment (mposolda) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/OIDCAdvancedRequestParamsTest.java]: Thanks for catching this. I've removed those lines. I've used separate commit, just so you can easily re-review that just this was changed. I plan to squash the commits before merging this.

**Comment by mposolda** [comment]
@tnorimat Thanks for the review and for doublecheck the FAPI and OIDC testsuite! I've addressed your comment in the OIDCAdvancedRequestParamsTest. Can you please re-review if looks ok now?

**Comment by tnorimat** [comment]
@mposolda Yes, I will re-review the PR.

**Review by tnorimat — APPROVED** [review]
LGTM

**Pull request merged**